TUMFARSynchrony/experimental-hub
# Trickle ICE

# Overview
This PR implements [trickle ICE](https://bloggeek.me/webrtcglossary/trickle-ice/) to all our WebRTC connections (Connections and Subconnections)

## Changes
The offer with the RTCSessionDescription is now send immediately, instead of waiting for all ICE candidates to be gathered. The candidates are then sent and added to the remote description afterwards.

### Connection
- Added new endpoint `/addIceCandidate`
- Candidates are associated with the correct connection using a unique id. It's generated on the client side and sent with the initial offer (`/offer)` as well as all candidates

### Subconnection
- Added new API call `ADD_ICE_CANDIDATE`
- Candidates are associated with the correct subconnection using the subconnection id, which is already present

### ICE Servers
Added more STUN and TURN servers to default config. This should make connection more reliable when using weird network configurations (e.g. dorm network)

### Wiki Changes
- [Connection Protocol](https://github.com/TUMFARSynchrony/experimental-hub/wiki/_compare/2952a036d58430bc37d88677cdf93ce94acb740a)
- [Connection Protocol Diagrams](https://github.com/TUMFARSynchrony/experimental-hub/wiki/Connection-Protocol/_compare/227c61cb04fd22d1c015e1d505a79d101d88c43c)
- [Data Types](https://github.com/TUMFARSynchrony/experimental-hub/wiki/_compare/a35508c58a018745fe0976f03359bb175d533f0a)

## How to test?
1. Make sure `REACT_APP_ICE_SERVERS=...` is not commented out in `.env`. You can also add more STUN/TURN servers if you like. (Don't forget to `npm run build`)
2. Create experiment and try to connect with different network configurations

It should work regardless of your network config and should be much quicker than before (if ICE servers were enabled).

Closes #162
